src/gpg-list-options.c

```diff
@@ -0,0 +1,147 @@
+#include "gpg-common.h"
+#include <string.h>
+#include <assert.h>
+#include <inttypes.h>
+#include <stdlib.h>
+
+struct lexbuf {
+    uint8_t *untrusted_cursor;
+};
+
+static bool buf_consume(struct lexbuf *const buf, uint8_t const expected)
+{
+    bool ret = buf->untrusted_cursor[0] == expected;
+    buf->untrusted_cursor += ret;
+    return ret;
+}
+
+static bool is_delim(uint8_t const untrusted_c)
+{
+    return untrusted_c == ' ' || untrusted_c == ',';
+}
+
+/* Skip spaces and commas */
+static void buf_consume_delims(struct lexbuf *const buf)
+{
+    while (is_delim(buf->untrusted_cursor[0]))
+        buf->untrusted_cursor++;
+}
+
+static void consume_subpacket_number(struct lexbuf *const buf)
+{
+    char *endptr = NULL;
+    long untrusted_subpacket_number = strtol((const char *)buf->untrusted_cursor, &endptr, 10);
+    if (endptr == NULL)
+        abort();
+    if ((const uint8_t *)endptr == buf->untrusted_cursor)
+        errx(1, "Invalid character in subpacket number list");
+    if (untrusted_subpacket_number < 1 || untrusted_subpacket_number > 127)
+        errx(1, "Subpacket number not valid (must be between 1 and 127 inclusive, got %ld)", untrusted_subpacket_number);
+    buf->untrusted_cursor = (uint8_t *)endptr;
+}
+
+static void consume_subpackets_argument(struct lexbuf *const buf)
+{
+    if (buf_consume(buf, '"')) {
+        if (!strchr((const char *)buf->untrusted_cursor, '"'))
+            errx(1, "Unterminated quoted option argument");
+        buf_consume_delims(buf);
+        while (!buf_consume(buf, '"')) {
+            consume_subpacket_number(buf);
+            buf_consume_delims(buf);
+        }
+    } else {
+        consume_subpacket_number(buf);
+    }
+}
+
+static size_t read_option(uint8_t *untrusted_option) {
+    for (size_t i = 0; ; ++i) {
+        switch (untrusted_option[i]) {
+        case '=':
+        case ',':
+        case ' ':
+        case '\0':
+            return i;
+        case 'A'...'Z': // GnuPG is case-insensitive, but this code is case-sensitive; casefold
+            untrusted_option[i] |= 0x20;
+            break;
+        default:
+            break;
+        }
+    }
+}
+
+static const struct listopt *find_option(const uint8_t *untrusted_option,
+                                         size_t len, bool *negated,
+                                         const struct listopt *options,
+                                         const char *msg)
+{
+    if ((*negated = (len >= 3 && !memcmp(untrusted_option, "no-", 3)))) {
+        untrusted_option += 3;
+        len -= 3;
+    }
+    for (; options->name; ++options) {
+        if (strlen(options->name) == len &&
+            memcmp(options->name, untrusted_option, len) == 0)
+            break;
+    }
+    if (!options->name)
+        errx(1, "Unknown %s option %.*s", msg, (int)len, untrusted_option);
```

Comment:
You dropped quotes around the option name. I'm fine with that (it's now consistent with other messages), but you need to update `test_080_option_parser`